File: javatests/com/google/protobuf/contrib/j2cl/integration/FloatFieldsTest.java
/*
 * Copyright 2019 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */
package com.google.protobuf.contrib.j2cl.integration;

import static com.google.common.truth.Truth.assertThat;
import static org.junit.Assert.assertThrows;

import com.google.protobuf.contrib.j2cl.protos.Accessor.TestProto;
import java.util.Arrays;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;

@RunWith(JUnit4.class)
public class FloatFieldsTest {
  private static Float PROTO_DEFAULT_FLOAT = 1.35f;
  private static Float DELTA = 0.00001f;

  @Test
  public void testOptionalFieldNoDefault_defaultInstance() {
    assertThat(TestProto.getDefaultInstance().getOptionalFloat()).isWithin(DELTA).of(0);
    assertThat(TestProto.getDefaultInstance().hasOptionalFloat()).isFalse();
  }

  @Test
  public void testOptionalFieldNoDefault_setValue() {
    TestProto.Builder builder = TestProto.newBuilder().setOptionalFloat(89.64f);
    assertThat(builder.hasOptionalFloat()).isTrue();
    assertThat(builder.getOptionalFloat()).isWithin(DELTA).of(89.64f);

    TestProto proto = builder.build();
    assertThat(proto.hasOptionalFloat()).isTrue();
    assertThat(proto.getOptionalFloat()).isWithin(DELTA).of(89.64f);
  }

  @Test
  public void testOptionalFieldNoDefault_setDefault() {
    TestProto.Builder builder = TestProto.newBuilder().setOptionalFloat(0f);
    assertThat(builder.hasOptionalFloat()).isTrue();
    assertThat(builder.getOptionalFloat()).isWithin(DELTA).of(0f);

    TestProto proto = builder.build();
    assertThat(proto.hasOptionalFloat()).isTrue();
    assertThat(proto.getOptionalFloat()).isWithin(DELTA).of(0f);
  }

  @Test
  public void testOptionalFieldNoDefault_clear() {
    TestProto startProto = TestProto.newBuilder().setOptionalFloat(1f).build();

    // Test clear
    TestProto.Builder builder = startProto.toBuilder();
    assertThat(builder.hasOptionalFloat()).isTrue();
    builder.clearOptionalFloat();
    assertThat(builder.hasOptionalFloat()).isFalse();

    TestProto proto = builder.build();
    assertThat(proto.hasOptionalFloat()).isFalse();
  }

  @Test
  public void testOptionalFieldWithDefault_defaultInstance() {
    assertThat(TestProto.getDefaultInstance().getOptionalFloatWithDefault())
        .isWithin(DELTA)
        .of(PROTO_DEFAULT_FLOAT);
    assertThat(TestProto.getDefaultInstance().hasOptionalFloatWithDefault()).isFalse();
  }

  @Test
  public void testOptionalFieldWithDefault_setDefaultValue() {
    TestProto.Builder builder =
        TestProto.newBuilder().setOptionalFloatWithDefault(PROTO_DEFAULT_FLOAT);
    assertThat(builder.hasOptionalFloatWithDefault()).isTrue();
    assertThat(builder.getOptionalFloatWithDefault()).isWithin(DELTA).of(PROTO_DEFAULT_FLOAT);

    TestProto proto = builder.build();
    assertThat(proto.hasOptionalFloatWithDefault()).isTrue();
    assertThat(proto.getOptionalFloatWithDefault()).isWithin(DELTA).of(PROTO_DEFAULT_FLOAT);
  }

  @Test
  public void testOptionalFieldWithDefault_setValue() {
    TestProto.Builder builder = TestProto.newBuilder().setOptionalFloatWithDefault(246f);
    assertThat(builder.hasOptionalFloatWithDefault()).isTrue();
    assertThat(builder.getOptionalFloatWithDefault()).isWithin(DELTA).of(246f);

    TestProto proto = builder.build();
    assertThat(proto.hasOptionalFloatWithDefault()).isTrue();
    assertThat(proto.getOptionalFloatWithDefault()).isWithin(DELTA).of(246f);
  }

  @Test
  public void testRepeatedField_defaultInstance() {
    assertThat(TestProto.getDefaultInstance().getRepeatedFloatCount()).isEqualTo(0);
    assertThat(TestProto.newBuilder().build().getRepeatedFloatCount()).isEqualTo(0);

    assertThrows(Exception.class, () -> TestProto.newBuilder().build().getRepeatedFloat(0));
  }

  @Test
  public void testRepeatedField_add() {
    TestProto.Builder builder =
        TestProto.newBuilder().addRepeatedFloat(1.324f).addRepeatedFloat(24f);
    assertThat(Arrays.asList(1.324f, 24f))
        .containsExactlyElementsIn(builder.getRepeatedFloatList())
        .inOrder();

    TestProto proto = builder.build();
    assertThat(Arrays.asList(1.324f, 24f))
        .containsExactlyElementsIn(proto.getRepeatedFloatList())
        .inOrder();
  }

  @Test
  public void testRepeatedField_addAll() {
    TestProto startProto =
        TestProto.newBuilder().addRepeatedFloat(1.324f).addRepeatedFloat(24f).build();

    TestProto.Builder builder =
        startProto.toBuilder().addAllRepeatedFloat(Arrays.asList(-1f, -2f, -3f));
    assertThat(Arrays.asList(1.324f, 24f, -1f, -2f, -3f))
        .containsExactlyElementsIn(builder.getRepeatedFloatList())
        .inOrder();

    TestProto proto = builder.build();
    assertThat(Arrays.asList(1.324f, 24f, -1f, -2f, -3f))
        .containsExactlyElementsIn(proto.getRepeatedFloatList())
        .inOrder();
  }

  @Test
  public void testRepeatedField_set() {
    TestProto startProto =
        TestProto.newBuilder()
            .addRepeatedFloat(1.324f)
            .addRepeatedFloat(24f)
            .addRepeatedFloat(-1f)
            .addRepeatedFloat(-2f)
            .addRepeatedFloat(-3f)
            .build();

    TestProto.Builder builder = startProto.toBuilder().setRepeatedFloat(2, 333f);
    assertThat(Arrays.asList(1.324f, 24f, 333f, -2f, -3f))
        .containsExactlyElementsIn(builder.getRepeatedFloatList())
        .inOrder();

    TestProto proto = builder.build();
    assertThat(Arrays.asList(1.324f, 24f, 333f, -2f, -3f))
        .containsExactlyElementsIn(proto.getRepeatedFloatList())
        .inOrder();
  }

  @Test
  public void testRepeatedField_getAndCount() {
    TestProto.Builder builder =
        TestProto.newBuilder()
            .addRepeatedFloat(1.324f)
            .addRepeatedFloat(24f)
            .addRepeatedFloat(-1f)
            .addRepeatedFloat(-2f)
            .addRepeatedFloat(-3f);

    assertThat(builder.getRepeatedFloatCount()).isEqualTo(5);
    assertThat(builder.getRepeatedFloat(0)).isWithin(DELTA).of(1.324f);
    assertThat(builder.getRepeatedFloat(1)).isWithin(DELTA).of(24f);
    assertThat(builder.getRepeatedFloat(2)).isWithin(DELTA).of(-1f);
    assertThat(builder.getRepeatedFloat(3)).isWithin(DELTA).of(-2f);
    assertThat(builder.getRepeatedFloat(4)).isWithin(DELTA).of(-3f);
    if (InternalChecks.isCheckIndex()) {
      assertThrows(Exception.class, () -> builder.getRepeatedFloat(5));
    } else {
      assertThat(builder.getRepeatedFloat(5)).isNaN();
    }

    TestProto proto = builder.build();
    assertThat(proto.getRepeatedFloat(0)).isWithin(DELTA).of(1.324f);
    assertThat(proto.getRepeatedFloat(1)).isWithin(DELTA).of(24f);
    assertThat(proto.getRepeatedFloat(2)).isWithin(DELTA).of(-1f);
    assertThat(proto.getRepeatedFloat(3)).isWithin(DELTA).of(-2f);
    assertThat(proto.getRepeatedFloat(4)).isWithin(DELTA).of(-3f);
    if (InternalChecks.isCheckIndex()) {
      assertThrows(Exception.class, () -> proto.getRepeatedFloat(5));
    } else {
      assertThat(proto.getRepeatedFloat(5)).isNaN();
    }
  }

  @Test
  public void testRepeatedField_clear() {
    TestProto startProto =
        TestProto.newBuilder()
            .addRepeatedFloat(1.324f)
            .addRepeatedFloat(24f)
            .addRepeatedFloat(-1f)
            .addRepeatedFloat(-2f)
            .addRepeatedFloat(-3f)
            .build();

    TestProto.Builder builder = startProto.toBuilder().clearRepeatedFloat();
    assertThat(builder.getRepeatedFloatCount()).isEqualTo(0);
    assertThat(builder.getRepeatedFloatList()).isEmpty();

    TestProto proto = builder.build();
    assertThat(proto.getRepeatedFloatCount()).isEqualTo(0);
    assertThat(proto.getRepeatedFloatCount()).isEqualTo(0);
  }
}

File: javatests/com/google/protobuf/contrib/immutablejs/integration/float_fields_test.js
// Copyright 2019 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

goog.module('proto.im.integration.FloatFieldsTest');
goog.setTestOnly();


const ListView = goog.require('proto.im.ListView');
const TestProto = goog.require('improto.protobuf.contrib.immutablejs.protos.TestProto');
const TestProto3 = goog.require('improto.protobuf.contrib.immutablejs.protos.TestProto3');
const testSuite = goog.require('goog.testing.testSuite');
const {assertEqualsForProto, assertValueIsCleared, assertValueIsSet} = goog.require('proto.im.proto_asserts');
const {isCheckIndex} = goog.require('proto.im.internal.internalChecks');


const PROTO_DEFAULT_VALUE = 1.35;

class FloatFieldsTest {
  testOptionalFieldNoDefault_defaultInstance() {
    assertEqualsForProto(0, TestProto.newBuilder().build().getOptionalFloat());
    assertEqualsForProto(
        false, TestProto.newBuilder().build().hasOptionalFloat());
  }

  testOptionalFieldNoDefault_setValue() {
    const builder = TestProto.newBuilder().setOptionalFloat(89.64);
    assertEqualsForProto(true, builder.hasOptionalFloat());
    assertEqualsForProto(89.64, builder.getOptionalFloat());

    const proto = builder.build();
    assertEqualsForProto(true, proto.hasOptionalFloat());
    assertEqualsForProto(89.64, proto.getOptionalFloat());
  }

  testOptionalFieldNoDefault_setDefaultValue() {
    const builder = TestProto.newBuilder().setOptionalFloat(0);
    assertEqualsForProto(true, builder.hasOptionalFloat());
    assertEqualsForProto(0, builder.getOptionalFloat());

    const proto = builder.build();
    assertEqualsForProto(true, proto.hasOptionalFloat());
    assertEqualsForProto(0, proto.getOptionalFloat());
  }

  testOptionalFieldNoDefault_clear() {
    const startProto = TestProto.newBuilder().setOptionalFloat(0).build();
    const builder = startProto.toBuilder();
    builder.clearOptionalFloat();
    assertEqualsForProto(false, builder.hasOptionalFloat());

    const proto = builder.build();
    assertEqualsForProto(false, proto.hasOptionalFloat());
  }

  testOptionalFieldWithDefault_defaultInstance() {
    assertEqualsForProto(
        PROTO_DEFAULT_VALUE,
        TestProto.newBuilder().build().getOptionalFloatWithDefault());
    assertEqualsForProto(
        false, TestProto.newBuilder().build().hasOptionalFloatWithDefault());
  }

  testOptionalFieldWithDefault_setDefaultValue() {
    const builder =
        TestProto.newBuilder().setOptionalFloatWithDefault(PROTO_DEFAULT_VALUE);
    assertEqualsForProto(true, builder.hasOptionalFloatWithDefault());
    assertEqualsForProto(
        PROTO_DEFAULT_VALUE, builder.getOptionalFloatWithDefault());

    const proto = builder.build();
    assertEqualsForProto(true, proto.hasOptionalFloatWithDefault());
    assertEqualsForProto(
        PROTO_DEFAULT_VALUE, proto.getOptionalFloatWithDefault());
  }

  testOptionalFieldWithDefault_setValue() {
    const builder = TestProto.newBuilder().setOptionalFloatWithDefault(135);
    assertEqualsForProto(true, builder.hasOptionalFloatWithDefault());
    assertEqualsForProto(135, builder.getOptionalFloatWithDefault());

    const proto = builder.build();
    assertEqualsForProto(true, proto.hasOptionalFloatWithDefault());
    assertEqualsForProto(135, proto.getOptionalFloatWithDefault());
  }

  testFieldWithDefault_serialization() {
    assertValueIsSet(TestProto.newBuilder().setOptionalFloat(0).build());

    const proto3builder = TestProto3.newBuilder();
    assertValueIsCleared(proto3builder.setProto3Float(0).build());
    assertValueIsSet(proto3builder.setProto3Float(1).build());
  }

  testRepeatedField_defaultInstance() {
    assertEqualsForProto(
        0, TestProto.newBuilder().build().getRepeatedFloatCount());
    assertEqualsForProto(
        0, TestProto.newBuilder().build().getRepeatedFloatList().size());
    if (isCheckIndex()) {
      assertThrows(() => TestProto.newBuilder().build().getRepeatedFloat(0));
    } else {
      assertUndefined(TestProto.newBuilder().build().getRepeatedFloat(0));
    }
  }

  testRepeatedField_add() {
    const builder =
        TestProto.newBuilder().addRepeatedFloat(1.324).addRepeatedFloat(24);
    assertEqualsForProto(2, builder.getRepeatedFloatCount());
    assertEqualsForProto([1.324, 24], builder.getRepeatedFloatList().toArray());


    const proto = builder.build();
    assertEqualsForProto(2, proto.getRepeatedFloatCount());
    assertEqualsForProto([1.324, 24], proto.getRepeatedFloatList().toArray());
  }

  testRepeatedField_addAll() {
    const startProto = TestProto.newBuilder()
                           .addRepeatedFloat(1.324)
                           .addRepeatedFloat(24)
                           .build();

    const builder = startProto.toBuilder().addAllRepeatedFloat(
        ListView.copyOf([-1, -2, -3]));
    assertEqualsForProto(
        [1.324, 24, -1, -2, -3], builder.getRepeatedFloatList().toArray());

    const proto = builder.build();
    assertEqualsForProto(
        [1.324, 24, -1, -2, -3], proto.getRepeatedFloatList().toArray());
  }

  testRepeatedField_set() {
    const startProto = TestProto.newBuilder()
                           .addRepeatedFloat(1.324)
                           .addRepeatedFloat(24)
                           .addRepeatedFloat(-1)
                           .addRepeatedFloat(-2)
                           .addRepeatedFloat(-3)
                           .build();

    const builder = startProto.toBuilder().setRepeatedFloat(2, 333);
    assertEqualsForProto(
        [1.324, 24, 333, -2, -3], builder.getRepeatedFloatList().toArray());

    const proto = builder.build();
    assertEqualsForProto(
        [1.324, 24, 333, -2, -3], proto.getRepeatedFloatList().toArray());
  }

  testRepeatedField_getAndCount() {
    const builder = TestProto.newBuilder()
                        .addRepeatedFloat(1.324)
                        .addRepeatedFloat(24)
                        .addRepeatedFloat(-1)
                        .addRepeatedFloat(-2)
                        .addRepeatedFloat(-3);


    assertEqualsForProto(5, builder.getRepeatedFloatCount());
    assertEqualsForProto(1.324, builder.getRepeatedFloat(0));
    assertEqualsForProto(24, builder.getRepeatedFloat(1));
    assertEqualsForProto(-1, builder.getRepeatedFloat(2));
    assertEqualsForProto(-2, builder.getRepeatedFloat(3));
    assertEqualsForProto(-3, builder.getRepeatedFloat(4));
    if (isCheckIndex()) {
      assertThrows(() => builder.getRepeatedFloat(5));
    } else {
      assertUndefined(builder.getRepeatedFloat(5));
    }

    const proto = builder.build();
    assertEqualsForProto(5, proto.getRepeatedFloatCount());

    assertEqualsForProto(1.324, proto.getRepeatedFloat(0));
    assertEqualsForProto(24, proto.getRepeatedFloat(1));
    assertEqualsForProto(-1, proto.getRepeatedFloat(2));
    assertEqualsForProto(-2, proto.getRepeatedFloat(3));
    assertEqualsForProto(-3, proto.getRepeatedFloat(4));
    if (isCheckIndex()) {
      assertThrows(() => proto.getRepeatedFloat(5));
    } else {
      assertUndefined(proto.getRepeatedFloat(5));
    }
  }

  testRepeatedField_clear() {
    const startProto = TestProto.newBuilder()
                           .addRepeatedFloat(1.324)
                           .addRepeatedFloat(24)
                           .addRepeatedFloat(-1)
                           .addRepeatedFloat(-2)
                           .addRepeatedFloat(-3)
                           .build();

    const builder = startProto.toBuilder().clearRepeatedFloat();
    assertEqualsForProto(0, builder.getRepeatedFloatCount());
    assertEqualsForProto(0, builder.getRepeatedFloatList().size());


    const proto = builder.build();
    assertEqualsForProto(0, proto.getRepeatedFloatCount());
    assertEqualsForProto(0, proto.getRepeatedFloatList().size());
  }
}

testSuite(new FloatFieldsTest());
